feat: hide full screen live chat button

OxrxL opened the pull request:
https://github.com/revanced/revanced-integrations/pull/189

OxrxL commented:
I preferred to open this PR to prevent me from forgetting that I created this patch...xD

OxrxL commented:
Rebased and tested again👍

OxrxL commented:
@oSumAtrIX

oSumAtrIX commented on src/main/kotlin/app/revanced/patches/youtube/layout/comments/bytecode/patch/CommentsPatch.kt:
```diff
@@ -63,6 +67,94 @@ class CommentsPatch : BytecodePatch(
+            listOf(LiveChatFullscreenButtonVisibilityFingerprint.result!!.classDef)
+        )
+
+        for (i in 0..setVisibilityAnchorFingerprints.size) {
```
Instead of creating an iterating index, simply iterate over each fingerprint.

OxrxL replied:
Ok...If for you is good, can I copy the same addinstructions() method for each fingerprint?

oSumAtrIX commented on src/main/kotlin/app/revanced/patches/youtube/layout/comments/bytecode/patch/CommentsPatch.kt:
```diff
@@ -63,6 +67,94 @@ class CommentsPatch : BytecodePatch(
+            with (setVisibilityAnchorFingerprints[i].result!!) {
+                val setVisibilityAnchorMethod = mutableMethod
+
+                fun visibilityPatchBuilder(viewRegister: Int, visibilityRegister: Int) = """
```
The name is confusing, the method is not building a patch.

oSumAtrIX commented on src/main/kotlin/app/revanced/patches/youtube/layout/comments/bytecode/patch/CommentsPatch.kt:
```diff
@@ -63,6 +67,94 @@ class CommentsPatch : BytecodePatch(
+        )
+
+        val setVisibilityAnchorFingerprints = listOf(
+            object : MethodFingerprint(
+                opcodes = listOf(
+                    Opcode.INVOKE_VIRTUAL,
+                    Opcode.CONST_4,
+                    Opcode.INVOKE_DIRECT,
+                    Opcode.IGET_OBJECT,
+                    Opcode.IGET_OBJECT,
+                    Opcode.NEW_INSTANCE,
+                    Opcode.INVOKE_DIRECT,
+                    Opcode.INVOKE_INTERFACE,
+                    Opcode.RETURN_VOID,
+                    Opcode.IGET_WIDE,
+                    Opcode.IGET_OBJECT,
+                    Opcode.INVOKE_VIRTUAL,
+                )
+            ) {},
+            object : MethodFingerprint(
+                "V", AccessFlags.PUBLIC or AccessFlags.FINAL, null, listOf(
+                    Opcode.IGET_OBJECT,
+                    Opcode.IGET,
+                    Opcode.INVOKE_VIRTUAL,
+                )
+            ) {},
+            object : MethodFingerprint(
+                "V", AccessFlags.PUBLIC or AccessFlags.FINAL, null, listOf(
+                    Opcode.IGET_OBJECT,
+                    Opcode.CONST_4,
+                    Opcode.INVOKE_VIRTUAL,
+                )
+            ) {}
```
It is suggested to create classes for each fingerprint out of convention, please do it here too.

OxrxL replied:
Can I placed these fingerprints in dedicated files, inside the fingerprints folder?

oSumAtrIX commented on src/main/kotlin/app/revanced/patches/youtube/layout/comments/bytecode/patch/CommentsPatch.kt:
```diff
@@ -63,6 +67,94 @@ class CommentsPatch : BytecodePatch(
+
+        liveChatFullscreenButtonMethod.addInstruction(
+            liveChatButtonIndex + 1, """
+                sput v${(liveChatFullscreenButtonMethod.instruction(liveChatButtonIndex) as OneRegisterInstruction).registerA}, Lapp/revanced/integrations/patches/HideLiveChatFullScreenButtonPatch;->fullScreenLiveChatButtonId:I
```
There are multiple very long and convoluted lines of duplicate code containing the full descriptor to methods from the integrations. Create variables for the descriptors to deduplicate.

oSumAtrIX commented on src/main/kotlin/app/revanced/patches/youtube/layout/comments/bytecode/patch/CommentsPatch.kt:
```diff
@@ -63,6 +67,94 @@ class CommentsPatch : BytecodePatch(
             """
         )
 
+    // hide fullscreen live chat button
```
Your comments are indented wrongly, please format your code.

oSumAtrIX commented on src/main/kotlin/app/revanced/patches/youtube/layout/comments/bytecode/patch/CommentsPatch.kt:
```diff
@@ -63,6 +67,94 @@ class CommentsPatch : BytecodePatch(
+                """
+
+                val invokeVirtualEndIndex =
+                    scanResult.patternScanResult!!.endIndex
```
In a lot of places you use the endIndex, create a method which returns a tuple from a given fingerprint which you can then deconstruct when saving the return value to a variable.

OxrxL replied:
For example?

OxrxL commented:
Let me know if I misunderstood a request.

OxrxL commented:
Applied some changes, in order:

- Converted the previous multiple fingerprints based code, on a direct class search based code (by the name of class, obtained from the single LiveChatFullscreenResource fingerprint).
- Removed the double class resolve by source id, for LiveChatFullscreenResource fingerprint.
- Added a companion method to get instruction by index (used multiple times).

oSumAtrIX reviewed (changes requested):
The current patch is too complex for its purpose. Please find a different, more readable or conscise implementation. Fingerprints can help with finding offsets.

OxrxL commented:
Closed until someone helps me implement this patch. I'm tired of looking for solutions that, at the end, are not good.
